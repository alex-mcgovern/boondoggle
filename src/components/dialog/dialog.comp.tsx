import { Slot } from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use_click_outside";
import { useForwardRef } from "../../hooks/use_forward_ref";
import { useOpenDialogWithKeyboard } from "../../hooks/use_open_dialog_with_keyboard";
import { a11yFocus } from "../../styles/common/a11y.css";
import { animateAppear } from "../../styles/common/animations.css";
import { Box } from "../box";
import { getDialogStyles } from "./dialog.styles.css";

import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./dialog.styles.css";
import type { ReactNode } from "react";

export type DialogProps = BoxProps & {
  children?: ReactNode;
  className?: string;
  inert?: boolean;
  isOpen?: boolean;
  onIsOpenChange?: (isOpen?: boolean) => void;
  placement?: DialogPlacementEnum;
  preventOpenOnKeydown?: boolean;
  triggerNode?: ReactNode;
  wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
  (
    {
      children,
      className: userClassName,
      isOpen: controlledIsOpen,
      onIsOpenChange,
      preventOpenOnKeydown = false,
      triggerNode,
      placement,
      wrapperProps,
      ...rest
    }: DialogProps,
    ref
  ) => {
    const dialogRef = useForwardRef<HTMLDialogElement>(ref);

    const triggerRef = createRef<HTMLElement>();

    /**
     * Callback for when the input is clicked or focused.
     */
    const toggleIsOpen = useCallback(() => {
      if (!dialogRef.current) {
        return undefined;
      }

      return (
        dialogRef.current?.open
          ? dialogRef.current?.close()
          : dialogRef.current?.show(),
        onIsOpenChange?.(dialogRef.current?.open)
      );
    }, [dialogRef, onIsOpenChange]);

    /**
     * When input is focused and user presses enter, open the date picker.
     */
    useOpenDialogWithKeyboard<HTMLElement>({
      callback: toggleIsOpen,
      dialogRef,
      preventOpenOnKeydown,
      triggerRef,
    });

    /**
     * Handle click outside dialog
     */
    useClickOutside<HTMLDialogElement, HTMLElement>({
      callback: () => {
        return dialogRef.current?.close();
      },
      contentRef: dialogRef,
      triggerRef,
    });

    return (
      <Box position="relative" {...wrapperProps}>
        <Slot onClick={toggleIsOpen} ref={triggerRef}>
          {triggerNode}
        </Slot>
        <Box
          {...rest}
          as="dialog"
          open={controlledIsOpen}
          ref={dialogRef}
          className={clsx(
            animateAppear,
            userClassName,
            a11yFocus,
            getDialogStyles({ placement })
          )}
        >
          {children}
        </Box>
      </Box>
    );
  }
);
