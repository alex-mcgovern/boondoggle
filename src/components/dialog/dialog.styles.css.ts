import { keyframes, style } from "@vanilla-extract/css";

import { vars } from "../../styles/theme.css";
import { createAccessibleTransition } from "../../styles/utils/create_accessible_transition";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

const overlayShow = keyframes({
  "0%": { opacity: 0 },
  "100%": { opacity: 1 },
});

const contentShow = keyframes({
  "0%": { opacity: 0, transform: "translate(-50%, -48%) scale(.96)" },
  "100%": { opacity: 1, transform: "translate(-50%, -50%) scale(1)" },
});

export const dialogOverlay = style([
  {
    background: `#00000080`,
    position: "fixed",
    inset: 0,
    zIndex: 99999,
  },
  createAccessibleTransition({
    animation: `${overlayShow} ${vars.transitionDuration.medium} cubic-bezier(0.16, 1, 0.3, 1) forwards`,
  }),
]);

export const dialogContent = style([
  getSprinkles({
    display: "block",
    isolation: "isolate",
    overflow: "hidden",
    textDecoration: "none",
    background: "tint_default",
    border: "border_default",
    borderRadius: "md",
    boxShadow: "lg",
    width: {
      mobile: "gridSpan10",
      tablet: "gridSpan8",
      desktop: "gridSpan8",
    },
    maxWidth: "gridSpan10",
  }),
  {
    zIndex: 99999,
    position: "fixed",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    selectors: {
      "&:focus": { outline: "none" },
    },
  },
  createAccessibleTransition({
    animation: `${contentShow} ${vars.transitionDuration.long} ease forwards`,
  }),
]);

export const dialogInner = style([
  getSprinkles({
    maxHeight: "90vh",
    padding: "spacing2",
  }),
  {
    overflowY: "auto",
  },
]);
