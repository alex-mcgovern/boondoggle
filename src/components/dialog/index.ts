export { Dialog } from "./dialog.component";
export type { DialogProps } from "./dialog.component";
