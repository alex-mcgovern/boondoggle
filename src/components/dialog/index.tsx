import * as RadixSlot from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use-click-outside";
import { useOpenDialogWithKeyboard } from "../../hooks/use-open-dialog-with-keyboard";
import { useForwardRef } from "../../hooks/use-forward-ref";
import { Box } from "../box";
import { getDialogStyles } from "./styles.css";

import type { ReactNode } from "react";
import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./styles.css";

export type DialogProps = BoxProps & {
	/**
	 * Children to render inside the dialog component.
	 */
	children?: ReactNode;

	/**
	 * Optional CSS class name applied to the outer HTML dialog element.
	 */
	className?: string;

	/**
	 * Whether the dialog is "inert" (can't be focused, can't be clicked.)
	 */
	inert?: boolean;

	/**
	 * Controls the dialog's open state, making it a controlled element. Use alongside `onIsOpenChange`.
	 */
	isOpen?: boolean;

	/**
	 * Callback when the dialog's open state is changed. Use alongside `isOpen`.
	 */
	onIsOpenChange?: (isOpen?: boolean) => void;

	/**
	 * Controls the interaction mode for the dialog trigger node.
	 */
	openOn?: "click" | "hover";

	/**
	 * Controls the dialog's placement relative to the trigger node.
	 */
	placement?: DialogPlacementEnum;

	/**
	 * Whether to open the dialog on an enter keypress while the trigger node is focused.
	 */
	preventOpenOnKeydown?: boolean;

	/**
	 * The react node to act as the trigger for the dialog.
	 */
	triggerNode?: ReactNode;

	/**
	 * `BoxProps` to pass to the outer element that wraps the dialog and it's trigger.
	 */
	wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
	(
		{
			children,
			className: userClassName,
			isOpen: controlledIsOpen,
			onIsOpenChange,
			openOn = "click",
			placement,
			preventOpenOnKeydown = false,
			triggerNode,
			wrapperProps,
			...rest
		}: DialogProps,
		ref,
	) => {
		const dialogRef = useForwardRef<HTMLDialogElement>(ref);

		const triggerRef = createRef<HTMLElement>();

		const toggleIsOpen = useCallback(() => {
			if (!dialogRef.current) {
				return;
			}

			if (dialogRef.current?.open) {
				dialogRef.current?.close();
			}

			if (!dialogRef.current?.open) {
				dialogRef.current?.show();
			}

			onIsOpenChange?.(dialogRef.current?.open);
		}, [dialogRef, onIsOpenChange]);

		useOpenDialogWithKeyboard<HTMLElement>({
			callback: toggleIsOpen,
			dialogRef,
			preventOpenOnKeydown,
			triggerRef,
		});

		useClickOutside<HTMLDialogElement, HTMLElement>({
			callback: () => {
				return dialogRef.current?.close();
			},
			contentRef: dialogRef,
			triggerRef,
		});

		return (
			<Box position="relative" {...wrapperProps}>
				<Box background="transparent">
					<RadixSlot.Slot
						onClick={toggleIsOpen}
						onMouseEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onMouseLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.close();
								  }
								: undefined
						}
						onPointerEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerOver={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						ref={triggerRef}
					>
						{triggerNode}
					</RadixSlot.Slot>
				</Box>
				<Box
					{...rest}
					as="dialog"
					className={clsx(
						userClassName,
						getDialogStyles({ placement }),
					)}
					open={controlledIsOpen}
					ref={dialogRef}
				>
					<Box>{children}</Box>
				</Box>
			</Box>
		);
	},
);
