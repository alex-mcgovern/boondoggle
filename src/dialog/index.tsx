import type { ComponentProps, ReactNode } from "react";
import type { DialogProps as RACDialogProps } from "react-aria-components";

import { faTimes } from "@fortawesome/pro-solid-svg-icons/faTimes";
import clsx from "clsx";
import {
    Dialog as RACDialog,
    DialogTrigger as RACDialogTrigger,
    Heading as RACHeading,
    Modal as RACModal,
    ModalOverlay as RACModalOverlay,
} from "react-aria-components";

import type { ColorOverlay } from "../index.css";

import { Button } from "../button";
import { css } from "../css/index.css";
import { Icon } from "../icon";
import {
    dialogCSS,
    dialogContentCSS,
    dialogFooterCSS,
    dialogHeaderCSS,
    dialogModalCSS,
    dialogTitleCSS,
    modalCSS,
    modalOverlayCSS,
} from "./styles.css";

export type DialogProps = RACDialogProps;

export function Dialog(props: DialogProps) {
    return (
        <RACDialog
            {...props}
            className={clsx(props.className, dialogCSS)}
        />
    );
}

export function DialogHeader({
    close,
    title,
}: {
    close: () => void;
    title: string;
}) {
    return (
        <header className={dialogHeaderCSS}>
            <RACHeading
                className={dialogTitleCSS}
                slot="title"
            >
                {title}
            </RACHeading>

            <Button
                appearance="ghost"
                aria-label="Close"
                className={css({
                    marginLeft: "auto",
                })}
                name="close"
                onPress={close}
                size="square_sm"
                type="button"
            >
                <Icon icon={faTimes} />
            </Button>
        </header>
    );
}

/**
 * Wrapper to render scrollable content within the dialog.
 */
export function ScrollableDialogContent({ children }: { children: ReactNode }) {
    return <div className={dialogContentCSS}>{children}</div>;
}

/**
 * Wrapper to pin content to the bottom of the dialog.
 */
export function DialogFooter({ children }: { children: ReactNode }) {
    return <footer className={dialogFooterCSS}>{children}</footer>;
}

export function DialogOld({
    buttonProps,
    children,
    colorOverlay,
    dialogTriggerProps,
    modalOverlayProps,
    modalProps,
    width = "sm",
}: {
    buttonProps?: ComponentProps<typeof Button>;
    children: ComponentProps<typeof RACDialog>["children"];
    /**
     * Color overlay, used to convey semantic meaning.
     * -   `amber`: To indicate caution, or warning actions
     * -   `blue`: To indicate information, or neutral actions
     * -   `green`: To indicate success, or confirmatory actions
     * -   `grey`: When a subtle, neutral color is needed
     * -   `red`: To warn of potentially destructive actions
     * -   `default`: To reset the color overlay if inherited from a parent
     */
    colorOverlay?: ColorOverlay;
    dialogTriggerProps?: Omit<
        ComponentProps<typeof RACDialogTrigger>,
        "children"
    >;
    modalOverlayProps?: Omit<
        ComponentProps<typeof RACModalOverlay>,
        "className"
    >;
    modalProps?: Omit<ComponentProps<typeof RACModal>, "className">;
    width?: "lg" | "sm";
}) {
    return (
        <RACDialogTrigger {...dialogTriggerProps}>
            {buttonProps ? <Button {...buttonProps} /> : null}
            <RACModalOverlay
                className={modalOverlayCSS}
                {...modalOverlayProps}
            >
                <RACModal
                    className={modalCSS({ colorOverlay, width })}
                    {...modalProps}
                >
                    <Dialog className={dialogModalCSS}>{children}</Dialog>
                </RACModal>
            </RACModalOverlay>
        </RACDialogTrigger>
    );
}
