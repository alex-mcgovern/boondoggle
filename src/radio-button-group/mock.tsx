import type * as React from "react";

import type { RadioButton } from "../radio-button";

import { LOREM } from "../../mocks/LOREM.mock";

export const RADIO_BUTTONS_MOCK: Array<
	React.ComponentProps<typeof RadioButton>
> = [
	{
		description: LOREM.text_xxs,
		name: "radio_item_1",
		title: "Radio item 1",
		value: "radio_item_1",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_2",
		title: "Radio item 2",
		value: "radio_item_2",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_3",
		title: "Radio item 3",
		value: "radio_item_3",
	},
];
