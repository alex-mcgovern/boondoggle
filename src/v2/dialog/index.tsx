import { faTimes } from "@fortawesome/pro-solid-svg-icons";
import {
	Dialog as ReactAriaDialog,
	DialogTrigger as ReactAriaDialogTrigger,
	Heading as ReactAriaHeading,
	Modal as ReactAriaModal,
	ModalOverlay as ReactAriaModalOverlay,
} from "react-aria-components";
import { ColorOverlay } from "../../index.css";
import { sprinkles } from "../../v1/sprinkles/index.css";
import { Button } from "../../button";
import { Icon } from "../icon";
import {
	dialogCSS,
	dialogContentCSS,
	dialogFooterCSS,
	dialogHeaderCSS,
	dialogTitleCSS,
	modalCSS,
	modalOverlayCSS,
} from "./styles.css";

/**
 * Wrapper to render the dialog header.
 */
export const V2DialogHeader = ({
	close,
	title,
}: { close: () => void; title: string }) => {
	return (
		<header className={dialogHeaderCSS}>
			<ReactAriaHeading slot="title" className={dialogTitleCSS}>
				{title}
			</ReactAriaHeading>

			<Button
				appearance="ghost"
				aria-label="Close"
				className={sprinkles({
					marginLeft: "auto",
				})}
				name="close"
				onPress={close}
				size="square_sm"
				type="button"
			>
				<Icon icon={faTimes} />
			</Button>
		</header>
	);
};

/**
 * Wrapper to render scrollable content within the dialog.
 */
export const V2ScrollableDialogContent = ({
	children,
}: { children: React.ReactNode }) => {
	return <div className={dialogContentCSS}>{children}</div>;
};

/**
 * Wrapper to pin content to the bottom of the dialog.
 */
export const V2DialogFooter = ({ children }: { children: React.ReactNode }) => {
	return <footer className={dialogFooterCSS}>{children}</footer>;
};

export const V2Dialog = ({
	buttonProps,
	colorOverlay,
	dialogTriggerProps,
	width = "sm",
	children,
	modalProps,
	modalOverlayProps,
}: {
	buttonProps?: React.ComponentProps<typeof Button>;
	colorOverlay?: ColorOverlay;
	dialogTriggerProps?: Omit<
		React.ComponentProps<typeof ReactAriaDialogTrigger>,
		"children"
	>;
	width?: "sm" | "lg";
	children: React.ComponentProps<typeof ReactAriaDialog>["children"];
	modalProps?: Omit<React.ComponentProps<typeof ReactAriaModal>, "className">;
	modalOverlayProps?: Omit<
		React.ComponentProps<typeof ReactAriaModalOverlay>,
		"className"
	>;
}) => {
	return (
		<ReactAriaDialogTrigger {...dialogTriggerProps}>
			{buttonProps ? <Button {...buttonProps} /> : null}
			<ReactAriaModalOverlay
				className={modalOverlayCSS}
				{...modalOverlayProps}
			>
				<ReactAriaModal
					className={modalCSS({ colorOverlay, width })}
					{...modalProps}
				>
					<ReactAriaDialog className={dialogCSS}>
						{children}
					</ReactAriaDialog>
				</ReactAriaModal>
			</ReactAriaModalOverlay>
		</ReactAriaDialogTrigger>
	);
};
