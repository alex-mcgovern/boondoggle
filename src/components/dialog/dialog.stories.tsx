import React from "react";

import { Dialog as StoryComponent } from ".";
import { WithDarkLightMode } from "../../../.storybook/with_dark_light_mode";
import { LOREM } from "../../../mocks/LOREM.mock";
import { Button } from "../button";

import type { DialogProps as StoryComponentProps } from ".";
import type { StoryFn, StoryObj } from "@storybook/react";

export default {
  title: "React components/Dialog",
  component: StoryComponent,
};

const Template: StoryFn<StoryComponentProps> = ({
  ...rest
}: StoryComponentProps) => {
  return <StoryComponent {...rest} />;
};

export const Default: StoryObj<StoryComponentProps> = {
  render: Template,
  name: "default",
  parameters: {
    layout: "fullscreen",
  },
  decorators: [WithDarkLightMode],
  args: {
    triggerNode: <Button name="Dialog button">Open dialog</Button>,
    children: LOREM.text_xxl,
    title: LOREM.text_xxs,
  },
};
