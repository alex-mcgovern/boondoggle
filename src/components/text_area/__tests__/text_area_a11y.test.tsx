/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react";

import { TextArea } from "..";
import { LOREM } from "../../../../mocks/LOREM.mock";
import "../../../../test/mocked_dependencies/dialog.mock";

import type { TextAreaProps } from "..";

const PROPS: TextAreaProps = {
    id: LOREM.id(),
    label: LOREM.label(),
    name: LOREM.name(),
    placeholder: LOREM.placeholder(),
};

const renderComponent = ({ ...props }: TextAreaProps) => {
    return render(<TextArea {...props} />);
};

describe("<TextArea />", () => {
    test("should assign name to the element", () => {
        const { getByRole } = renderComponent(PROPS);

        expect((getByRole("textbox") as HTMLTextAreaElement).name).toBe(PROPS.name);
    });

    test("should be labelled", () => {
        const { container } = renderComponent(PROPS);

        const label = container.querySelector("label");

        expect(label).not.toBeNull();

        expect(label?.getAttribute("for")).toBe(PROPS.id);

        expect(label?.textContent).toBe(PROPS.label);

        // expect(
        //   getByLabelText(PROPS.label as string, { selector: "text_area" })
        // ).not.toBeNull();
    });

    test("should match snapshot", () => {
        const { getByRole } = renderComponent(PROPS);

        expect(getByRole("textbox")).toMatchSnapshot();
    });
});
