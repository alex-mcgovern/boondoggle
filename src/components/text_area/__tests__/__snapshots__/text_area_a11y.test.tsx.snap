// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should match snapshot 1`] = `
<textarea
  aria-label="Label ipsum"
  class="styles_getTextAreaStyles__jyp1w4 styles__jyp1w3 get_sprinkles_width_100%_mobile__eiij7bgn a11y_a11yDisabled__r8he5t1 a11y_a11yFocus__r8he5t2 styles_getTextAreaStyles_size_md__jyp1w6 styles_size_md__jyp1w1 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b21 get_sprinkles_lineHeight_body_md__eiij7b3f element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_spacing_1.5_mobile__eiij7bdz get_sprinkles_paddingRight_spacing_1.5_mobile__eiij7bew get_sprinkles_paddingTop_spacing_1_mobile__eiij7bfq get_sprinkles_paddingBottom_spacing_1_mobile__eiij7bcz"
  id="lorem-ipsum"
  name="Name ipsum"
  placeholder="Placeholder ipsum"
/>
`;
